eng/Get-ServiceModules.ps1

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env pwsh
+
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+<#
+    .Synopsis
+    Gets the modules that are built from the given service directory and SDK type.
+
+    .Parameter ServiceDirectory
+    Name of the service being built. For example, "eventhubs".
+    .Parameter SdkType
+    Type of SDK to find modules for.
+
+    .Example
+    .\eng\Get-ServiceModules "eventhubs" "client"
+    Will output "azure-messaging-eventhubs,azure-messaging-eventhubs-checkpointstore-blob" because
+    those are modules that are built by the 'client' SDK type.
+
+    .Example
+    .\eng\Get-ServiceModules "eventhubs" "data"
+    Will output "microsoft-azure-eventhubs,microsoft-azure-eventhubs-eph,microsoft-azure-eventhubs-extensions"
+    because those are modules that are built by the 'data' SDK type.
+#>
+[CmdletBinding()]
+param(
+    [Parameter(Mandatory = $true, Position = 0)]
+    [ValidateNotNullOrEmpty()]
+    [string]$ServiceDirectory,
+
+    [Parameter(Mandatory = $true, Position = 1)]
+    [ValidateSet("data", "client")]
+    [string]$SdkType
+)
+
+# By default stop for any error.
+if (!$PSBoundParameters.ContainsKey("ErrorAction")) {
+    $ErrorActionPreference = "Stop"
+}
+
+switch($SdkType) {
+    "client" {
+        $parentArtifactId = "azure-client-sdk-parent"
+        break
+    }
+    "data" {
+        $parentArtifactId = "azure-data-sdk-parent"
+        break
+    }
+    default {
+        throw "'$SdkType' is not supported."
+    }
+}
+
+# If it is not a module that is built by pom.client.xml, we should skip it.
+$clientModules = New-Object System.Collections.Generic.HashSet[string]
```

Comment:
This is going to be fragile. In the near future we are going to be removing the ```<module />``` elements from the parent POM which means you won't be able to rely on this.

Reply:
That's why it was trying to build azure-core-management, because it couldn't be found in pom.client.xml and azure-core's pom.service.xml is a reference to pom.client.xml. Any other suggestions?

Reply:
This is going to be messy :(

We kinda lean on the ```java - core -ci```/```java - core``` pipelines to build everything as a bit of a broad integration test, which is why we reference ```pom.client.xml``` as a proxy for building those things that depend on core in any way. It's a bit of a hack.

The more correct thing to do is basically have another script which scans everything under SDK looking for modules that take a dependency on the modules under ```sdk/[servicedir]``` and also build them. So for example storage would be unable to check anything in that broke event hubs (its not just about core).

The problem is that without restructuring our parent POMs (or at least using pure aggregators) we aren't going to be able to do the right thing here. Which means that this code will need to be reworked pretty soon.

The right thing (but big effort thing) to do here is:

1. Place a true aggregate POM in ```./pom.xml``` which then references each service POM.
2. Change build scripts to build against that root POM using the arguments that you are generating here.
3. Change your script to do two scans. The first scan scans ```sdk/[service]/*``` for the complete list of POMs under a service directory. Then scan everything under ```sdk/*``` and build a list of targets that depend directly on the modules in the target ```sdk/[service]``` folder. Then taking that list, and the target directory modules create the ```-pl``` switch.

This would net us a few benefits. First, you wouldn't inadvertently build storage for event hubs. However, storage would build event hubs, because it depends on storage. For live testing we probably wouldn't include downstream dependents in the test.

Sadly the correct thing to do here seems like a lot of work.

Reply:
> This is going to be fragile. In the near future we are going to be removing the `<module />` elements from the parent POM which means you won't be able to rely on this.

@mitchdenny There is a lot of tooling which relies on inspecting the parent pom entry to figure out what track pom belongs to which track.